### PR TITLE
WIP: Ditching MailboxProcessor

### PR DESCRIPTION
**Pull request by et1975**
@MangelMaxime @vbfox @alfonsogarciacaro @Zaid-Ajaj 

Caveat: .NET/WPF will need to augment the  `dispatch` with synchronization on the UI thread, cause the new dispatch loop is not thread-safe.

Let's have some eyeballs on this - it seems to work interactively, but..
My elm-performance fork is not happy with it - the new TODO textbox appears to be filled out, but I get no  message about it, so it tries adding an empty TODO and `update` ends up ignoring it.

### CODE REVIEW

**Comment by et1975** [comment]
Published as 3.0.0-beta-1

**Comment by MangelMaxime** [comment]
TBH I don't really understand the code, so I can't help much for giving a review. 🤷‍♂️ 

Perhaps adding some comments to help explaining what is doing what can help if others want to look at the code. ?

**Comment by forki** [comment]
what wwas the reason for getting rid of it?

**Comment by MangelMaxime** [comment]
@forki This is related to
- https://github.com/elmish/elmish/issues/157
- https://github.com/elmish/elmish/issues/130

Small resume, it's for benefiting of React optimization out of the box. Currently we don't have them because we get a new referece on `dispatch` on each "loop" and react consider that it's a different function and so send an update for all our event handlers in the view.

**Comment by forki** [comment]
So for .NET we could keep it and use compiler directive?

Am Mo., 12. Nov. 2018, 09:05 hat Maxime Mangel <notifications@github.com>
geschrieben:

**Comment by et1975** [comment]
@forki the main issue is that currently MBP as implemented by fable does not guarantee FIFO for commands. As Maxime points out, there are benefits to React, but WPF would benefit from the dispatch loop running on the UI thread: https://github.com/elmish/Elmish.WPF/issues/40

`setState` can already be overriden to augment `dispatch` for the views, but I'll probably just introduce a `Program` property `syncDispatch` to do that, so that it works for commands as well. 

@MangelMaxime the ring is just a fancy queue (because there's no F# queue I'm having to implement one, so why not make it a fancy one ;)).

File: netstandard/Elmish.fsproj [hunk 1]
```diff
@@ -8,7 +8,9 @@
   <ItemGroup>
     <Compile Include="../src/prelude.fs" />
     <Compile Include="../src/cmd.fs" />
+    <Compile Include="../src/ring.fs" />
     <Compile Include="../src/program.fs" />
+    <Compile Include="meta.fs" />
```
Comment (et1975) [inline netstandard/Elmish.fsproj]: I can probably make do w/o this, we'll see if just referencing the ring source from the test project makes more sense that this jumping through the hoops.

File: src/program.fs [hunk 2]
```diff
@@ -87,31 +87,35 @@ module Program =
...
-                        with ex ->
-                            program.onError ("Unable to process a message:", ex)
-                            state
-                    return! loop newState
-                }
-            loop model
-        )
-        program.setState model inbox.Post
+        let mutable rb = RingBuffer 10
+        let mutable reentered = false
```
Comment (et1975) [inline src/program.fs]: flag indicating that dispatch is entered recursively via a cmd that was processed as part of earlier entry in the stack.

File: src/program.fs [hunk 3]
```diff
@@ -87,31 +87,35 @@ module Program =
...
+                rb.Push msg
+            else
+                reentered <- true
+                let mutable nextMsg = Some msg
+                while Option.isSome nextMsg do // I wish we could write Rust-like `while (Some msg = nextMsg) do`
+                    let msg = nextMsg.Value
+                    try
+                        let (model',cmd') = program.update msg state
+                        program.setState model' dispatch
+                        cmd' |> Cmd.exec dispatch
```
Comment (et1975) [inline src/program.fs]: This is where re-entry can happen and this is where Fable currently fails to FIFO.

File: tests/RingBuffer.fs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+let internal applyRBOps (ops:Op list) (rb:int RingBuffer) =
+    ops
+    |> List.map (function 
+                 | Pop -> 
+                   Poped <| rb.Pop()
+                 | Push v -> 
+                   rb.Push v
+                   Pushed)
+                   
+[<Property>]
```
Comment (et1975) [inline tests/RingBuffer.fs]: This is FsCheck test, it generates 100 test cases by default, I think.

File: tests/RingBuffer.fs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+let actsLikeAQueue (ops:Op list) =  
+    let q = Queue()
+    let rb = RingBuffer 10
+    (q |> applyQOps ops) =! (rb |> applyRBOps ops)
```
Comment (et1975) [inline tests/RingBuffer.fs]: Here we're asserting that the same operations on the queue and the buffer result in the same outcome.

**Comment by et1975** [comment]
cc @cmeeren 

In .NET land you'll need to override `syncDispatch` on the program instance and set it to `Dispatcher.Invoke` or some such, because new dispatch loop is not thread safe.

**Comment by drk-mtr** [comment]
@et1975 this is a bit of a long shot but I'm trying to work out why I'm seeing errors when running this: https://github.com/DieselMeister/Terminal.Gui.Elmish/tree/master/src/Terminal.Gui.Elmish/ElmishImplementation and I've just come across your comment.

I'm trying to determine whether trying your "override `syncDispatch`" suggestion might help, but I don't really understand how I'd do this. Would this involve adding another pipe when starting the program? e.g.

```fsharp
Program.mkProgram init update view
|> Program.withSyncDispatch xxx
|> Program.run
```

I wasn't able to find any reference to `Dispatcher` or `Invoke` within the Elmish repo so I've hit a bit of a dead end i.e. I've got no idea what the `xxx` would be (if I'm even on the right lines above that is).

If this involves actually implementing a dispatcher please ignore my query since that's probably beyond my capabilites at the moment!

**Comment by et1975** [comment]
Hey @drk-mtr,

Since you have a copy of elmish in your code base you could just revert the
dispatch loop implementation to the one from the v2 branch:
https://github.com/elmish/elmish/blob/v2.x/src/program.fs#L88.
Alternatively, the syncDispatch implementation just needs to make sure only
one dispatch call happens at a time, something like:

let queuedDispatch dispatch =
    let inbox = MailboxProcessor.Start(fun (mb:MailboxProcessor<'msg>) ->
        let rec loop () =
            async {
                let! msg = mb.Receive()
                dispatch msg
                return! loop ()
            }
        loop ()
    )
    inbox.Post

**Comment by bender2k14** [comment]
@drk-mtr, I implemented essentially the same thing as @et1975 in response to a similar question in #181.  (And now I cleaned up the code with @et1975's improvements, so it is exactly the same.)  See https://github.com/elmish/elmish/issues/181#issuecomment-530534973.

**Comment by drk-mtr** [comment]
Huge thanks for the suggestions - really appreciate it!

**Pull request merged**